### PR TITLE
[CPU] GatherElements, GatherND, GatherTree migration on nGraph.

### PR DESCRIPTION
**Pull request by nshchego**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp [hunk 1]
```diff
@@ -12,70 +14,81 @@ namespace InferenceEngine {
...
-                outputData->getTensorDesc().getLayoutByDims(outDims));
-        config.outConfs.push_back(outConfig);
+            const auto& dataDims = op->get_input_shape(dataIndex_);
+            const auto& indicesDims = op->get_input_shape(indicesIndex_);
+            if (dataDims.size() != indicesDims.size())
+                IE_THROW() << errorPrefix_ << " has invalid input shapes. Inputs 'Data' and 'Indices' must have equal ranks.";
 
-        config.dynBatchSupport = false;
+            Precision inDataPrecision = details::convertPrecision(op->get_input_element_type(dataIndex_));
+            if (inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I32>::value_type) &&
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp:52]: use one_of please

File: inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp [hunk 2]
```diff
@@ -12,70 +14,81 @@ namespace InferenceEngine {
...
+                    inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I16>::value_type) &&
+                    inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I8>::value_type)) {
+                IE_THROW() << errorPrefix_ << " has unsupported 'inputData' input precision: " << inDataPrecision;
+            }
+
+            Precision indicesPrecision = details::convertPrecision(op->get_input_element_type(indicesIndex_));
+            if (indicesPrecision != Precision::I32) {
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp:59]: \+ I64 for further

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp [hunk 3]
```diff
@@ -13,75 +15,84 @@ namespace InferenceEngine {
...
-        Precision dataPrecision = data->getTensorDesc().getPrecision();
-        if (dataPrecision.size() != sizeof(PrecisionTrait<Precision::I32>::value_type) &&
-                dataPrecision.size() != sizeof(PrecisionTrait<Precision::I16>::value_type) &&
-                dataPrecision.size() != sizeof(PrecisionTrait<Precision::I8>::value_type)) {
-            IE_THROW() << _errorPrefix << " has unsupported 'data' input precision: " << dataPrecision;
+    static bool isSupportedOperation(const std::shared_ptr<ngraph::Node>& op, std::string& errorMessage) noexcept {
+        try {
+            auto gatherElementsOp = ngraph::as_type_ptr<const ngraph::op::v5::GatherND>(op);
+            if (!gatherElementsOp) {
+                errorMessage = "Node is not an instance of the GatherElements operation from operation set v6.";
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp:26]:
```suggestion
                errorMessage = "Node is not an instance of the GatherND operation from operation set v5.";
```

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp [hunk 4]
```diff
@@ -13,75 +15,84 @@ namespace InferenceEngine {
...
-            IE_THROW() << _errorPrefix << " has invalid batch_dims attribute: " << _batchDims;
+            if (op->get_input_size() != 2 || op->get_output_size() != 1)
+                IE_THROW() << _errorPrefix << " has invalid number of input/output edges.";
 
-        _batchNum = 1lu;
-        for (size_t i = 0; i < _batchDims; i++) {
-            _batchNum *= indicesDims[i];
-        }
+            Precision inDataPrecision = details::convertPrecision(op->get_input_element_type(_dataIndex));
+            if (inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I32>::value_type) &&
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp:48]: one_of

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp [hunk 5]
```diff
@@ -13,75 +15,84 @@ namespace InferenceEngine {
...
+                    inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I8>::value_type)) {
+                IE_THROW() << _errorPrefix << " has unsupported 'data' input precision: " << inDataPrecision;
+            }
 
-        _sliceRank = indicesDims[indicesDims.size() - 1];
-        _dataRank = dataDims.size() - _batchDims;
-        if (_sliceRank > _dataRank)
-            IE_THROW() << _errorPrefix << " has invalid inputs shapes.";
+            Precision indicesPrecision = details::convertPrecision(op->get_input_element_type(_indicesIndex));
+            if (indicesPrecision != Precision::I32 &&
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp:55]: maybe better to use ngraph ```Type``` class method ```is_integral()```?

File: inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp (hunk 1 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp (hunk 2 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp (hunk 3 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp]: FIxed

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp (hunk 4 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp]: FIxed

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp (hunk 5 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp]: is_float() from IE was added instead

File: inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp [hunk 6]
```diff
@@ -12,70 +15,82 @@ namespace InferenceEngine {
...
+                    sizeof(PrecisionTrait<Precision::I8>::value_type))) {
+                IE_THROW() << errorPrefix_ << " has unsupported 'inputData' input precision: " << inDataPrecision;
+            }
 
-        const auto& outDims = outputData->getTensorDesc().getDims();
-        outConfig.desc = TensorDesc(dataPrecision, outDims,
-                outputData->getTensorDesc().getLayoutByDims(outDims));
-        config.outConfs.push_back(outConfig);
+            Precision indicesPrecision = details::convertPrecision(op->get_input_element_type(indicesIndex_));
+            if (indicesPrecision.is_float()) {
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp:61]: according to spec only i32 and i64 precision supported

File: inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp (hunk 5 above)
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp]: what if indicesPrecision will be BOOL?

Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp]: Fixed

File: inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp (hunk 6 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/gather_elements.cpp]: Fixed

**Review by mandrono — APPROVED** [review]

**Pull request merged**